release-23.2: server/license: Make enterprise license check a no-op

spilchen opened the pull request:
Backport 1/1 commits from #131943.

/cc @cockroachdb/release

---

With the deprecation of the core license, there’s no longer a need to check whether enterprise features are enabled—all features are now always enabled. If a license policy violation occurs, such as an expired license, we will throttle connections, but all enterprise features will remain functional. I’ve left the API for checking if enterprise features are enabled in the code, but these functions are now no-ops and can be removed in the future.

I also had to adjust some tests that were explicitly disabling enterprise features, as that’s no longer possible. Some tests were rewritten to fit the new model, while others were removed entirely because they were no longer relevant.

Note: A few tests in `pkg/sql` had to be updated due to the changes in enterprise checks. With all enterprise features enabled, the default isolation level has also changed. Any test that explicitly disabled leases (using TestingDisableTableLeases) needed to apply this setting _after_ starting up the test server. Applying it earlier, as was originally done, caused the test server to fail during startup.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes: CRDB-41998
Release note: none
Release justification: This work is part of the CockroachDB core deprecation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132342)
<!-- Reviewable:end -->

spilchen committed:
server/license: Make enterprise license check a no-op

With the deprecation of the core license, there’s no longer a need to
check whether enterprise features are enabled—all features are now
always enabled. If a license policy violation occurs, such as an expired
license, we will throttle connections, but all enterprise features will
remain functional. I’ve left the API for checking if enterprise features
are enabled in the code, but these functions are now no-ops and can be
removed in the future.

I also had to adjust some tests that were explicitly disabling
enterprise features, as that’s no longer possible. Some tests were
rewritten to fit the new model, while others were removed entirely
because they were no longer relevant.

Note: A few tests in pkg/sql had to be updated due to the changes in
enterprise checks. With all enterprise features enabled, the default
isolation level has also changed. Any test that explicitly disabled
leases (using TestingDisableTableLeases) needed to apply this setting
after starting up the test server. Applying it earlier, as was
originally done, caused the test server to fail during startup.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes: CRDB-41998
Release note: none
